### DISCUSSION TITLE
Ferramentas para trabalhar no Windows

### DISCUSSION THREAD

**Discussion by pedrohenri** [frontendbr/forum#112] [Discussão]
Para alguns, encontrar as melhores ferramentas para realizar determinadas tarefas no Windows pode ser um desafio, principalmente quando o trabalho envolve a command line.

Esta discussão tem como objetivo primário receber contribuições sobre quais ferramentas você usa para tornar o desenvolvimento no Windows produtivo.

Em resumo que ferramentas vocês usam para editar código, usar a command line, setar um servidor local, fazer ssh, etc.

Existe uma outra discussão aberta sobre se o SO realmente tem limitações ou não para o desenvolvimento. Vamos manter as coisas organizadas e discutir aqui apenas quais são as melhores ferramentas para quem trabalha usando Windows.

**Comment by karlprieb** [comment]
Já tentei trabalhar no Windows diversas vezes, mas nunca fui feliz hehehe O filename char limit é complicado, ainda mais quando trabalha com Node. Terminal fraco, etc. Talvez com a vinda do Bash fique bem melhor. Estou aguardando ansiosamente o Ubuntu on Windows para testar :)

**Comment by leocavalcante** [comment]
1. **Editar código**
   É praticamente irrelevante, todos os editores e IDEs do mercado dão suporte pro Windows. Eu pessoalmente gosto do [Atom](https://atom.io/).
2. **Setar um servidor**
   Também é "meio" irrelevante porque depende do servidor. Se for Apache eu recomendo o [Apache Lounge](https://www.apachelounge.com/).
3. **Fazer SSH**
   É mais simples do que parece. Se você instalar a distribuição convencional do [Git](https://git-scm.com/) pro Windows você já tem SSH no Git Bash e se durante a instalação você permitir que o wizard ponha todos os comandos que vem com o Git no PATH (ultima opção com um warning em vermelho) você vai ter SSH em qualquer terminal. De qualquer forma, ainda tem Putty e WinSCP, mas pô, que dev não tem Git pra subir uns projetinhos open-source no Github?
4. **Terminal/Console**
   Tem um projeto muito f#da chamado [ConEmu](https://conemu.github.io/) que emula um console com abas, divisões e tudo mais de legal que tem nos terminais do Unix. Eu pessoal prefiro um superset do ConEmu chamado [Cmder](http://cmder.net/), basicamente: ele deixa o ConEmu mais foda ainda.
5. **Dica**
   Seta uma variável de ambiente chamada `HOME` apontando pro diretório do seu usuário, ex. `C:\Users\Leo` Alguns programas e scripts esperam por ela.
6. **MinGW & Cygwin**
   [MinGW](http://www.mingw.org/) é um subset do Unix compilado pra rodar no Windows, ele é normalmente usado de forma incorporada por quem tá distribuindo algum software, exemplos são o próprio [Git](https://git-scm.com/), [DevKit](http://rubyinstaller.org/add-ons/devkit/) e [OCPWin](https://www.typerex.org/ocpwin.html)
   [Cygwin](https://www.cygwin.com/) é bem mais completo, ele tenta ser um próprio Linux no seu Windows e tem uma [lista extensa de pacotes GNU](https://cygwin.com/packages/package_list.html) pra rodar no Windão, `make`, `emacs`, `vim`, `vi`, `grep`, `perl`... quase um Linux mesmo.
7. **Ubuntu BASH no Windows**
   Pra quem esteve em outro planeta nos últimos dias: a Microsoft e a Canonical (empresa por trás do Ubuntu) fecharam uma pareceria pra trazer o Ubuntu pro Windows e entregar esse workflow que teria num Unix pro Windows, praticamente um Cygwin só que nativo, são os binários do Ubuntu rodando no Win. https://msdn.microsoft.com/en-us/commandline/wsl/about

**Comment by karlprieb** [comment]
Essa issue não pode ser mergeada com a #113 ?

**Comment by leocavalcante** [comment]
8.. **Too long filename**
O @karlprieb me lembrou de algo importante que todo mundo acaba topando ainda mais trabalhando com Node.js. A solução é o [`robocopy`](https://technet.microsoft.com/en-us/library/cc733145.aspx), já vem no Windows e em teoria serve "só" pra copiar arquivos só que com parametros certos você consegue copiar uma pasta vazia pra pasta que você quer excluir e acabar deletando o conteúdo dela depois só deletar ela vazia sem problemas. Eu criei um .bat que deixo num diretório do PATH chamado `rmforce.bat`

``` batch
@echo off
mkdir rmforce_placeholder
robocopy rmforce_placeholder %1 /purge
rmdir rmforce_placeholder
rmdir %1
```

Aí só chamar
`rmforce pasta_com_subniveis_grandes`

**Comment by karlprieb** [comment]
@leocavalcante Quando estava no Windows acabei instalando o rimraf (node package) e usava ele para deletar folders.

**Comment by felipefialho** [comment]
Do carario seu post @leocavalcante.

Não conhecia esse ConEmu, utilizo o Console2.

Acredita que o ConEmu é melhor nesse aspecto?

**Comment by karlprieb** [comment]
Só para avisar :) Saiu a primeira versao do Ubuntu on Windows :)

https://blogs.msdn.microsoft.com/commandline/2016/04/06/bash-on-ubuntu-on-windows-download-now-3/

Testarei! hehehe

**Comment by leocavalcante** [comment]
@LFeh, usei bastante o Console2, se não me engano ele é um superset do ConEmu igual o Cmder, mas migrei pro Cmder por ser um projeto mais ativo https://github.com/cmderdev/cmder O ultimo update do Console2 foi em 2013.

**Comment by victormiguez** [comment]
Tenho usado o cmder por mais de um ano, tem sido muito bom.
Recomendo bastante o uso!

**Comment by felipefialho** [comment]
Sim, também acho o Console2 abandonado pra cacete.

Vou dar uma olhada no Cmder então. 

---

Edit. Já tinha até estrelado o projeto, hehe

**Comment by schiehll** [comment]
pra contribuir...
https://github.com/dthree/cash

comandos do linux em es6, só instalar o package via npm e sair usando

**Comment by felipefialho** [comment]
Instalei o Cmder, vou passar a usar a partir de amanhã para testar.

Depois passo minhas experiencias.

**Comment by raphaklaus** [comment]
Já botei minha opinião numa issue parecida. Mas só gostaria de dizer que estou feliz de ver front-ends usando o Winzão. :)

Várias vezes me senti forever alone sendo o único non-mac da turminha. =p

**Comment by danguilherme** [comment]
Sempre usei Windows. Cmder salva vidas! <3

Já tive problemas sim, mas nenhum era impossível de ser solucionado. Ruby é
bem mais difícil de instalar no Windows, mas mesmo assim foi
inesperadamente simples - algumas coisinhas para configurar, mas é
rapidinho. Já devem ter sofrido bem mais antes de mim pra fazer de uma
maneira tão redonda como está agora.

Mas realmente, ele tem limitações (no nosso caso) e não é o foco dos
desenvolvedores Open Source. As soluções geralmente vêm primeiro pra Mac e
Linux, e geralmente alguém da comunidade arruma um jeito de funcionar no
Windows (mais comum quando falamos de ferramentas). Bom, não deixa de ser
da comunidade, mas demora mais pra sair.

**Comment by felipefialho** [comment]
Caras, só pra dar um feedback.

Estou usando o **Cmder**, é realmente bem melhor que o Console2. Já coloquei inclusive no contexto de menu para facilitar abrir direto no diretório que eu quero =)

Estou me familiarizando com os atalhos agora.

**Comment by dougaraujos** [comment]
Eu costumo investir muito em produtividade, principalmente trabalhando com Windows.
@leocavalcante valew pela ótima dica! Comecei a usar o **Cmder** e curti demais!

**Comment by vitor-mendrone-deel** [comment]
O Cmder é animal! Recomendo pra todos que usam Windows. Conheci faz alguns meses, dica do @gmoura

**Comment by lagden** [comment]
Esse é velho mais ajudou uma galerinha!!!
http://lagden.in/2014/05/29/instalando-o-git-node-js-ruby-sass-compass-bower-grunt-etc-no-windows/

**Souvenir** ![megaman](http://vignette1.wikia.nocookie.net/megaman/images/2/22/MM2-AirShooter-Sprite.png/revision/latest?cb=20101023180731)

**Comment by felipefialho** [comment]
Boa @lagden

**Comment by karlprieb** [comment]
Alguém ja testou o Ubuntu on Windows?

**Comment by felipefialho** [comment]
@karlprieb Não conheço, o que seria?

---

Alguém sabe um jeito abrir arquivos do Sketch no Windows, sem ser convertendo para SVG?

**Comment by karlprieb** [comment]
@felipefialho Foi anunciado no /build/ (evento da MS). Dá uma olhada https://blogs.windows.com/buildingapps/2016/03/30/run-bash-on-ubuntu-on-windows/

**Comment by felipefialho** [comment]
Ah sim, deve ser implementado em um próximo update né?

**Comment by karlprieb** [comment]
Aham... no verão deles agora vai sair stable... Atualmente esta no insider preview

**Comment by thiagopov** [comment]
power shell ta funfando bem no w10.

**Comment by lagden** [comment]
Eu rodei o Ubuntu no Windows em 2015, mas tinha que se inscrever no `Windows Insider Program`
Era muito limitado... o `Cygwin` é bem melhor

**Comment by carloshenriqueribeiro** [comment]
Faço parte do WIP e rodo o Ubuntu no W10. Tá bem melhor que no começo. 💪

Carlos Ribeiro

**Comment by onlurking** [comment]
Ruby no Windows é quase sempre uma experiência terrível, então pra remediar eu uso o [Cmder](http://cmder.net/) em conjunto com o [MSYS2](http://www.msys2.org/) - que é basicamente o Cygwin  com o gerenciador de pacotes do Arch Linux.

Também uso um Launcher chamado [Wox](http://www.getwox.com/) e o [Everything](https://www.voidtools.com/) pra melhorar minha produtividade em procurar arquivos e abrir aplicações, coisas que no Linux eu faço usando o `Rofi` como launcher e `ag` e `ANGRYsearch` pra procurar.

**Comment by fredmosc** [comment]
Atualmente em casa uso só Windows por motivos de computador compartilhado
kkkkk.
Uso docker para quase tudo qdo vou programar no windows, não quero outra
vida

Em qui, 30 de nov de 2017 às 09:40, Diogo Felix <notifications@github.com>
escreveu:

> Ruby no Windows é quase sempre uma experiência terrível, então pra
> remediar eu uso o Cmder <http://cmder.net/> em conjunto com o MSYS2
> <http://www.msys2.org/> - que é basicamente o Cygwin com o gerenciador de
> pacotes do Arch Linux.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/frontendbr/forum/issues/112#issuecomment-348164318>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJB79VtdU-lY3E5vWXwEaCeKrLMy9Wpnks5s7pQkgaJpZM4IBIhc>
> .
>
-- 
Att,

FM
<https://about.me/fredmosc?promo=email_sig&utm_source=product&utm_medium=email_sig&utm_campaign=edit_panel&utm_content=thumb>
Frederick Moschkowich
about.me/fredmosc
<https://about.me/fredmosc?promo=email_sig&utm_source=product&utm_medium=email_sig&utm_campaign=edit_panel&utm_content=thumb>

**Comment by felipefialho** [comment]
Aproveitando a deixa

### [Usando o Terminal do Linux no Windows](https://www.felipefialho.com/blog/2017/usando-o-terminal-do-linux-no-windows)

**Comment by lagden** [comment]
Concordo com o @fredmosc 

Docker é melhor solução nos últimos tempos...
Cultura DevOps...

**Comment by leocavalcante** [comment]
👀

**Comment by willianjusten** [comment]
Tá aí uma discussão antiga, mas que agora "amadureceu bem". Hoje o Windows com o WSL2 permitiu ter praticamente tudo que queríamos do Linux, mas dentro do Windows mesmo xD

Sobre o Terminal, eu uso o incrível [Windows Terminal](https://github.com/microsoft/terminal), que tem recebido várias atualizações sempre. E aí junto dele, já uso o [oh-my-zsh](https://ohmyz.sh/) com alguns plugins para autocomplete, ssh, nvm e etc.

Como editor, nem preciso dizer muito né, o [VS Code](https://code.visualstudio.com/) tem imperado.

Para trabalhar com banco de dados, eu tenho usado o [DBeaver](https://dbeaver.io/), consigo acessar MySQL, Postgres e outros, tudo de forma fácil.

Eu tenho um [mini-curso](https://www.youtube.com/watch?v=YcR8pKvjx44&list=PLlAbYrWSYTiOpefWtd6uvwgKT1R-94Zfd), onde eu explico como montar um sistema bem similar ao meu, talvez seja útil =)

**Reply by leocavalcante** [reply]
Pois é, acho que dos 7 itens que apontei lá em cima, só o 3 que continuou igual, o projeto Git-for-Windows evoluiu muito bem. Apesar de que tem gente que prefere até o Git usar do WSL.

**Reply by willianjusten** [reply]
Eu uso tudo tudo do WSL agora xD

**Reply by leocavalcante** [reply]
Se eu pegasse um Win do zero acho que provavelmente faria assim também rs Como já tenho um monte de coisa configurada, da preguiça de sair migrando tudo rsrs

**Comment by leocavalcante** [comment]
Dois items legais de mencionar também na thread como um todo é o https://chocolatey.org/ e o https://scoop.sh/ são basicamente o brew/apt/yum/apk do Windows.
